akka-stream/src/main/scala/akka/stream/ContextMapStrategy.scala

```diff
@@ -4,53 +4,35 @@
 
 package akka.stream
 
-object ContextMapStrategy {
-
-  final case class Iterate[In, Ctx, Out](
-      iterateFn: (In, Ctx, Out, Int, Boolean) => Ctx,
-      emptyFn: Option[(In, Ctx) => (Out, Ctx)] = None)
-
-  /**
-   * Passthrough the same context [[Ctx] of the input element [[In]] to all output elements [[Out]].
-   */
-  def same[In, Ctx, Out](): Iterate[In, Ctx, Out] = Iterate(
-    iterateFn = (_: In, inCtx: Ctx, _: Out, _: Int, _: Boolean) => inCtx
-  )
-
-  /**
-   * Transform the context of the first element given the input element [[In]] and context [[Ctx]].
-   */
-  def first[In, Ctx, Out](firstFn: (In, Ctx, Out) => Ctx): Iterate[In, Ctx, Out] = Iterate(
-    iterateFn = (in: In, inCtx: Ctx, out: Out, index: Int, _: Boolean) =>
-      if (index == 0) firstFn(in, inCtx, out)
-      else inCtx
-  )
+/**
+ * A ContextMapStrategy defines how the context parameter of a FlowWithContext is manipulated
+ * as various operations are performed in the flow
+ *
+ * @typeparam Ctx the context type
+ */
+class ContextMapStrategy[-Ctx]
 
+object ContextMapStrategy {
   /**
-   * Transform the context of the last element given the input element [[In]], context [[Ctx]], and the index
-   * of the element.
+   * Trait that is mixed into strategies that allow elements (including their offsets) to be filtered out of the stream.
    */
-  def last[In, Ctx, Out](fn: (In, Ctx, Out, Int) => Ctx): Iterate[In, Ctx, Out] = Iterate(
-    iterateFn = (in: In, inCtx: Ctx, out: Out, index: Int, hasNext: Boolean) =>
-      if (!hasNext) fn(in, inCtx, out, index)
-      else inCtx
-  )
+  trait Filtering[-Ctx] {
+    self: ContextMapStrategy[Ctx] =>
+  }
 
   /**
-   * Iterate over each element and transform the context given the element [[In]], context [[InCtx]],
-   * output element [[Out]], index, and if the there is a next output element [[Out]]. This strategy can be used to
-   * satisfy all 1:many use cases.
+   * Trait that is mixed into strategies that allow elements (including their offsets) to be reordered.
    */
-  def iterate[In, Ctx, Out](fn: (In, Ctx, Out, Int, Boolean) => Ctx): Iterate[In, Ctx, Out] = Iterate(fn)
+  trait Reordering[-Ctx] {
```

Comment:
```suggestion
  trait Unordered[-Ctx] {
```
`Reordering` sounded to me like a promise that elements would be reordered while `unordered` is both consistent with the operator name and sounds to me like "maybe ordered and maybe not".

Reply:
What would you think about `AllowsReordering`? A bit more verbose, but might make the meaning more clear?